src/main/scala/co/teapot/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -15,51 +15,77 @@
  *
- * When transforming a graph where nodeCount <= maxNodeId
+ * Graphs are converted to this format using MemoryMappedDirectedGraphConverter, then loaded
+  * using this class.  When transforming a graph where nodeCount <= maxNodeId
  * to this format, new nodes with no neighbors will be implicitly created. The binary
  * format is currently subject to change.
  */
 
-/* Storage format
+/* Internally, the graph is stored in a large binary file made of a header and a series of segments.
+Out-neighbor segment i contains the out-neighbors of all nodes whose id satisfies
+  id % segmentCount == i
+and similarly for in-neighbor segment i.  Each segment has a sequence of offsets, where the jth
+offset in out-neighbor segment i stores the offset (as an int relative to the start of that segment) for
+node
+  id = j * segmentCount + i
+where nodesPerSegment = nodeCount / segmentCount.  We require for simplicity that nodeCount be
+a multiple of segmentCount (by creating empty nodes if needed).
+
+Storage format
+Header:
 byteCount  data
-8          (reserved, later use for versioning or indicating undirected vs directed)
-8          n (i. e. the number of nodes).  Currently must be less than 2^31.
-8*(n+1)    Offsets into out-neighbor data. Index i (a Long) points to the out-neighbor data of
-           node i.  The out-neighbor data must be stored in sequential order by id, as the
-           outegree of node i is computed from the difference in offset between node i+1 and node i.
-           Index n is needed to compute the outdegree of node n - 1.
-8*(n+1)    Offsets into in-neighbor data (Longs) (Same interpretation as out-neighbor offsets)
-4*m        out-neighbor data
-4*m        in-neighbor data
+8          (version number, currently 1)
+4          n (nodeCount).  Must be a multiple of segmentCount.
+8          edgeCount
+4          k (segmentCount)
+8*(k + 1)  The absolute offset in bytes where each out-neighbor segment's data begins.  The "+ 1"
+           is for computing the size of the last segment.
+8*(k + 1)  (Similarly, for in-neighbors)
+
+Each out-neighbor segment:
+byteCount  data
+4*(n/k+1)  Offset in bytes into out-neighbor data relative to the start of this segment. Index j
+           plus the offset of this segment points to the out-neighbor data of node
+           id = (j * segmentCount + i).  The
+           out-neighbor data is be stored in increasing order by id, so the outegree of node
+           id is computed from the difference in offset between node id + segmentCount and node id.
+           One additional index is needed to compute the outdegree of the last node in this segment.
+4*(# edges in segment)        out-neighbor data
+
+Each in-neighbor segment is analogous to an out-neighbor segment.
  */
 class MemoryMappedDirectedGraph(file: File) extends DirectedGraph {
-  val data: MemoryMappedIntLongSource = new MemoryMappedIntLongSource(file)
+  val fileChannel = FileChannel.open(file.toPath, StandardOpenOption.READ)
```

Comment:
Why are we switching this? Is this an orthogonal issue? It doesn't seem like adding offset segments would require this to change, since everything is still a long or an int.